### ISSUE TITLE
Out of memory error

### ISSUE THREAD

**Issue by cgy0627** [bcgsc/LINKS#57]
Hello,

I have a contig file of a large size (~19 Gb) that I would like to scaffold using the PacBio data, but I keep getting the "Out of memory!" error. When I tried running it using only 500 reads from a total of 136,774,993 PacBio reads, it ran successfully, meaning the PacBio data seems to be too big to run LINKS. All my other options were left to default.

Is there a way to split up the PacBio data and run them on parallel to combine the results afterwards? What's the maximum size allowed for the PacBio input data?
Could I simply split it up into 100 files for instance and run them separately 100 times iteratively using the same bloom file?

Any help would be greatly appreciated!

Thank you,
Jenny

**Comment by warrenlr** [comment]
Hi Jenny,

One way to curb memory usage is to use larger values for the -t parameter.  
Increasing it will have for effect to explore a more restricted kmer pair space, while still reading-in all sequencing reads. This issue has been raised before and you may be able to find a guideline to choosing an initial value that works best with your system.

As a rule of thumb, you may first want to calculate the N50 length of your read set. This will give you an indication on the intervals to choose (ie. -d parameter). I would recommend you run LINKS iteratively where the output of first -d runs be used as input to the next ones.  As you increase -d, you may also decrease -t since less and less contigs will harbor kmer pairs at higher intervals (esp. if your initial assembly is fragmented).

So, maybe try:

-d 5000, -t 200 (if you run out of memory, increase -t further. If it runs on your system, consider halving the value or reducing it until it fits)
-d 10000, -t 150
-d 15000, -t 100
-d 20000, -t 80

Since I don't know anything about your data&system, difficult for me to estimate the parameters, but it is a good place to start.

Good luck!
René

**Comment by cgy0627** [comment]
Hi René,

Thank you for a quick reply!

The N50 length of my read set is 22,065 and I am trying out the options you have suggested. If I want to increase -t further, is there a specific proportion with which I should decrease -d?
Also, I couldn't find the github repo you mentioned as I am quite new to the system. Could you maybe post a link?

Thank you so much for your time.

Jenny

**Comment by warrenlr** [comment]
Hi Jenny,

```If I want to increase -t further, is there a specific proportion with which I should decrease -d```

-t is a window step parameter. When you set it to 100, for example, each kmer pairs -d XX bp apart will be extracted every -t bases. 

The combination of -d and -t that gives scaffolding depends on so many factors, which include but no limited to:
-read coverage
-read length
-read base quality
-genome size
-genome contiguity
-system specs (RAM but will be limited at 500GB-1TB even if system has more memory) 

What's more important is that you fit "-t" for a specific value of "-d" you wish to scaffold at (and you should always scaffold iteratively with shorter to longer distances in increasing order). And for your long reads (N50~20kbp), you can explore iterative runs at -d 5000, 10000, 15000, 20000, 25000, 30000 to begin. You can even go lower with -d, but keep in mind that it is only possible with such a large genome, if -t allows the extracted kmer pairs to fit in the RAM.

this is an example issue reported previously. Most issues relate to RAM limitations, which can usually be dealt with by increasing -t. You just need to troubleshoot and adjust it until it fits into memory.

https://github.com/bcgsc/LINKS/issues/55

**Comment by cgy0627** [comment]
Hi René,

I tried the iteration, but it wouldn't go past the first line of code. So just to see if the first "-t" option works I tried the following:
`LINKS -f contig.fa -s Pacbio.fof -b test -d 5000 -t 1000 -k 26`

After trying out 200 and 500, which both got killed during k-mer extraction stage, the option -t 1000 finally went past the k-mer extracting stage, but was killed during the scaffolding phase.

`Contigs (>= 500 bp) processed k=26:
85687

=>Writing Bloom filter to disk (test.bloom) : Fri Apr 16 19:49:54 KST 2021
Writing a 35520019112 byte filter to test.bloom on disk.

=>Reading long reads, building hash table : Fri Apr 16 19:50:17 KST 2021
Reads processed k=26, dist=5000, offset=0 nt, sliding step=1000 nt:

Reads processed from file 1/1, Pacbio.fasta.gz:
136774993
Extracted 5268408 26-mer pairs at -d 5000, from all 136774993 sequences provided in Pacbio.fof

Extracted 5268408 26-mer pairs overall. This is the set that will be used for scaffolding

=>Reading sequence contigs (to scaffold), tracking k-mer positions : Sat Apr 17 02:30:35 KST 2021
Contigs (>= 500 bp) processed k=26:
85687

=>Scaffolding initiated :
0-3.Links_k26.sh: line 4: 3650505 Killed                  LINKS -f contig.fa -s Pacbio.fof -b test -d 5000 -t 1000 -k 26`

Do you think I should increase the -t option further? I read in another post that -t 1000 was already a large enough value, and I couldn't find other posts related to this. Perhaps my contigs.fa (19 Gb) or Pacbio read data (bloomfilter is 35 Gb) are too large?

Thanks again for your help.

Jenny

**Comment by warrenlr** [comment]
Hi Jenny,

Coud you please confirm that the demo ran successfully? Just want to make sure there's no reason to doubt the install.

How much RAM does your system have?  Sounds like that is the limiting factor (because 5,268,408 kmer pairs extracted isn't that much).. 

We are able to scaffold spruce genomes (20Gb) with at least 20X ONT reads, and -t wasn't that high (150 - 200 max at lower intervals) but did require ~500-750 GB RAM

-t 1000 is quite high, but the only way to make this process complete is to increase further I am afraid (unless you find a system with more memory). You original idea of subsampling the reads may be another solution, but a higher -t is essentially having a similar effect, without necessarily biasing the sequences.. I still think you should try doubling -t until you get a run that completes.  Once you have that, you may start decreasing it to get the most kmer pairs that fit your system's memory.

Take care,
Rene

**Comment by warrenlr** [comment]
I should also add that once you have a Bloom filter built from your draft at a specific -k value (k26 in your case), you don't need to rebuild it each time you run LINKS, it will save you time.

You just pass it like so:

-r test.bloom

------------------------------------------

Also, related to building the Bloom filter, if your system isn't equipped with too much RAM, you may be able to reduce the size of the Bloom filter by increasing the FPR. This is done through option -p.  By default it is set at 0.1%, I wouldn't increase it more than 5% but you could do tests at -p 0.01 (1%) to begin.  To re-build the Bloom filter, run LINKS as you have been doing, but simply add -p 0.01 to your command. This will build a Bloom filter that should be smaller in size but still filter error kmers.

**Comment by cgy0627** [comment]
Hi René,

Thanks for your quick reply!

I ran the demo using `./runme_EcoliK12single.sh` and checked that it worked perfectly fine. Also, my system has 1.11 T of RAM, which seems to be enough, but maybe my Pacbio (70X) reads are too much? The reads have not been filtered in any way and I'm not too sure of the quality. The following is the information for my Pacbio data:
```
Number of sequences             136,774,993

Residue counts:
  Total                     1,759,578,234,841

Sequence lengths:
  Minimum                       50
  Maximum                      120,809
  Average                         12,864.77
  N50                                22,065
  N80                                12,240
  N90                                7,587
```

Since the scaffolding process only uses the extracted k-mer pairs from the Pacbio reads, so I guess it wouldn't matter too much as long as the k-mers have been extracted properly?

For now I will give -t 2000 and -p 0.01 a try like you suggested and will let you know!

Thanks,
Jenny

**Comment by warrenlr** [comment]
your system seem to have enough RAM..  unfortunately PERL isn't the best at memory management and we can't seem to harness enough for these runs. Yes, 70X is a lot, but again a large -t would have a somewhat similar effect as subsambling, without risking joining contigs from evidence of a single read.

I suggest you precede your LINKS command by:

```/usr/bin/time -v -o test.time ./LINKS -f contig.fa -s Pacbio.fof -b test -d 5000 -t 1000 -k 26```

and when/if it crashes, let's look at the maximum resident memory.

Also, from experience, with reads of that size, the most gains will be observed between intervals of 7500-15000 bp. 
So, for the sake of getting a run to complete, perhaps you could relaunch with:

```/usr/bin/time -v -o test_k26_d10kt1k.time ./LINKS -f contig.fa -s Pacbio.fof -b test -d 10000 -t 1000 -k 26 -r test.bloom```

**Comment by warrenlr** [comment]
this may be fixed by v2.0 (C++) version just released. Please reopen if it's still an issue. thx
